File: app/src/main/java/com/thegreek/niko/logviewer_for_openHAB/MainFragment.java
package com.thegreek.niko.logviewer_for_openHAB;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.fragment.app.Fragment;

import java.util.Objects;

import es.dmoral.toasty.Toasty;

public class MainFragment extends Fragment {

    private Button connectButton;
    private CheckBox connectCheck;
    private CheckBox hostnameIPAddressCheck;
    private CheckBox portCheck;
    private EditText hostnameIPAddress;
    private EditText port;
    private ImageButton hostnameIPAddressEdit;
    private ImageButton portEdit;
    private SharedPreferences mySPR;
    private SharedPreferences.Editor editor;
    private String hostnameIPAddressString;
    private String link;
    private TextView linkView;

    private int portInt;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View v = inflater.inflate(R.layout.fragment_main, container, false);

        hostnameIPAddress = v.findViewById(R.id.hostname_ip_address);
        hostnameIPAddressEdit = v.findViewById(R.id.hostname_ip_address_edit);
        hostnameIPAddressCheck = v.findViewById(R.id.hostname_ip_address_check);
        port = v.findViewById(R.id.port);
        portEdit = v.findViewById(R.id.port_edit);
        portCheck = v.findViewById(R.id.port_check);
        linkView = v.findViewById(R.id.link_view);
        connectButton = v.findViewById(R.id.connect_button);
        connectCheck = v.findViewById(R.id.connect_check);
        TextView endUserConsent = v.findViewById(R.id.end_user_consent);
        TextView credits_text = v.findViewById(R.id.credits);
        TextView versionView = v.findViewById(R.id.version_view);
        ImageView settings = v.findViewById(R.id.imageView);

        mySPR = v.getContext().getSharedPreferences("Safe", 0);
        editor = mySPR.edit();
        editor.apply();

        statusWiederherstellung();
        versionView.setText(BuildConfig.VERSION_NAME);

        settings.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int newOrientation = mySPR.getInt("orientation", 0) == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE ? ActivityInfo.SCREEN_ORIENTATION_PORTRAIT :
                        ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE;
                MainActivity.changeOrientation(Objects.requireNonNull(getActivity()), newOrientation);
                editor.putInt("orientation", newOrientation).apply();

                String orientation = mySPR.getInt("orientation", 0) == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE ? "landscape" : "portrait";
                Toasty.info(Objects.requireNonNull(getContext()), "Orientation changed to " + orientation, Toasty.LENGTH_SHORT).show();
            }
        });

        connectButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (!hostnameIPAddress.getText().toString().isEmpty() && !port.getText().toString().isEmpty()) {
                    if (linkView.getText().toString().isEmpty()) {
                        linkGenerierung();

                        hostnameIPAddress.setEnabled(false);
                        hostnameIPAddressEdit.setVisibility(View.VISIBLE);
                        port.setEnabled(false);
                        portEdit.setVisibility(View.VISIBLE);
                        hostnameIPAddressCheck.setVisibility(View.INVISIBLE);
                        portCheck.setVisibility(View.INVISIBLE);
                        connectCheck.setVisibility(View.VISIBLE);

                        if (hostnameIPAddressCheck.isChecked()) {
                            editor.putString("hostnameIPAddressString", hostnameIPAddressString);
                            editor.putBoolean("hostnameIPAddressCheck", true).apply();
                        } else {
                            editor.putString("hostnameIPAddressString", "");
                            editor.putBoolean("hostnameIPAddressCheck", false).apply();
                        }

                        if (portCheck.isChecked()) {
                            editor.putInt("portInt", portInt);
                            editor.putBoolean("portCheck", true).apply();
                        } else {
                            editor.putInt("portInt", 0);
                            editor.putBoolean("portCheck", false);
                            editor.apply();
                        }

                        editor.putString("link", link).apply();

                        connectButton.setText(getString(R.string.connect_button_2));
                    } else {
                        assert getFragmentManager() != null;
                        getFragmentManager().beginTransaction()
                                .replace(R.id.start, new WebViewFragment())
                                .addToBackStack(null)
                                .commit();

                        if (view != null) {
                            InputMethodManager imm = (InputMethodManager) view.getContext().getApplicationContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                            assert imm != null;
                            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
                        }

                        Toasty.info(v.getContext(), getString(R.string.connecting), Toasty.LENGTH_SHORT).show();
                    }
                } else {
                    Toasty.error(v.getContext(), getString(R.string.error_fill_out), Toasty.LENGTH_LONG).show();
                }
            }
        });

        hostnameIPAddressEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                hostnameIPAddress.setEnabled(true);
                hostnameIPAddressEdit.setVisibility(View.INVISIBLE);
                portEdit.setVisibility(View.INVISIBLE);
                hostnameIPAddressCheck.setVisibility(View.VISIBLE);
                connectCheck.setVisibility(View.INVISIBLE);

                linkView.setText("");
                connectButton.setText(getString(R.string.connect_button_1));
            }
        });

        portEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                port.setEnabled(true);
                hostnameIPAddressEdit.setVisibility(View.INVISIBLE);
                portEdit.setVisibility(View.INVISIBLE);
                portCheck.setVisibility(View.VISIBLE);
                connectCheck.setVisibility(View.INVISIBLE);

                linkView.setText("");
                connectButton.setText(getString(R.string.connect_button_1));
            }
        });

        endUserConsent.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                EndUserConsent2 endUserConsent2 = new EndUserConsent2();
                endUserConsent2.setCancelable(true);
                endUserConsent2.show(getChildFragmentManager(), "EndUserConsent2");
            }
        });

        credits_text.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent termsOfUseIntent = new Intent(getContext(), Credits.class);
                view.getContext().startActivity(termsOfUseIntent);
            }
        });

        connectCheck.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                editor.putBoolean("connectCheck", connectCheck.isChecked());
                editor.putBoolean("autoStart", connectCheck.isChecked()).apply();
            }
        });

        return v;
    }

    private void statusWiederherstellung() {
        MainActivity.changeOrientation(Objects.requireNonNull(getActivity()), mySPR.getInt("orientation", 0));

        connectCheck.setChecked(mySPR.getBoolean("connectCheck", false));

        if (mySPR.getBoolean("autoStart", false) && connectCheck.isChecked()) {
            assert getFragmentManager() != null;
            getFragmentManager().beginTransaction()
                    .replace(R.id.start, new WebViewFragment())
                    .addToBackStack(null)
                    .commit();

            Toasty.info(getActivity().getApplicationContext(), getString(R.string.connecting), Toasty.LENGTH_SHORT).show();

            return;
        }

        if (mySPR.getString("hostnameIPAddressString", "").equals("") || mySPR.getString("hostnameIPAddressString", "").equals("0")) {
            hostnameIPAddress.setText("");
            hostnameIPAddress.setEnabled(true);
        } else {
            hostnameIPAddress.setText(mySPR.getString("hostnameIPAddressString", "0"));
            hostnameIPAddressString = mySPR.getString("hostnameIPAddressString", "0");
            hostnameIPAddress.setEnabled(false);
        }

        hostnameIPAddressCheck.setChecked(mySPR.getBoolean("hostnameIPAddressCheck", true));

        if (mySPR.getInt("portInt", 0) == 0) {
            port.setText("");
            hostnameIPAddress.setEnabled(true);
        } else {
            port.setText(String.valueOf(mySPR.getInt("portInt", 0)));
            port.setEnabled(false);
            portInt = mySPR.getInt("portInt", 0);
        }

        portCheck.setChecked(mySPR.getBoolean("portCheck", true));

        if (!hostnameIPAddress.getText().toString().isEmpty() && !port.getText().toString().isEmpty()) {
            linkGenerierung();

            hostnameIPAddressEdit.setVisibility(View.VISIBLE);
            portEdit.setVisibility(View.VISIBLE);
            hostnameIPAddressCheck.setVisibility(View.INVISIBLE);
            portCheck.setVisibility(View.INVISIBLE);
            connectCheck.setVisibility(View.VISIBLE);

            connectButton.setText(getString(R.string.connect_button_2));
        }
    }

    private void linkGenerierung() {
        hostnameIPAddressString = hostnameIPAddress.getText().toString();
        portInt = Integer.parseInt(port.getText().toString());
        link = "http://" + hostnameIPAddress.getText().toString() + ":" + portInt;
        linkView.setText(link);
    }
}

File: app/src/main/java/com/cyb3rko/logviewerforopenhab/fragments/MainFragment.kt
package com.cyb3rko.logviewerforopenhab.fragments

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.core.view.get
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.cyb3rko.logviewerforopenhab.*
import com.cyb3rko.logviewerforopenhab.databinding.FragmentMainBinding
import com.google.android.material.button.MaterialButton

class MainFragment : Fragment() {
    private var _binding: FragmentMainBinding? = null
    // This property is only valid between onCreateView and onDestroyView.
    private val binding get() = _binding!!
    private lateinit var myContext: Context

    private lateinit var editor: SharedPreferences.Editor
    private lateinit var hostnameIPAddressString: String
    private lateinit var link: String
    private lateinit var mySPR: SharedPreferences
    private var portInt = 0

    override fun onAttach(context: Context) {
        super.onAttach(context)
        myContext = context
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentMainBinding.inflate(inflater, container, false)
        val root = binding.root

        mySPR = myContext.getSharedPreferences(SHARED_PREFERENCE, Context.MODE_PRIVATE)
        editor = mySPR.edit()
        editor.apply()

        setEditButtonClickListener()
        setConnectButtonClickListener(root)
        setConnectCheckClickListener()

        return root
    }

    override fun onStart() {
        super.onStart()
        statusRestoring()
    }

    private fun statusRestoring() {
        if (!mySPR.getBoolean(HTTPS_ACTIVATED, false)) {
            binding.httpToggles.check((binding.httpToggles[0] as MaterialButton).id)
        } else {
            binding.httpToggles.check((binding.httpToggles[1] as MaterialButton).id)
        }
        binding.connectCheck.isChecked = mySPR.getBoolean(CONNECT_CHECK, false)

        if (mySPR.getString(HOSTNAME_STRING, "") == "" || mySPR.getString(HOSTNAME_STRING, "") == "0") {
            binding.hostnameIpAddressText.setText("")
            binding.hostnameIpAddress.isEnabled = true
        } else {
            binding.hostnameIpAddressText.setText(mySPR.getString(HOSTNAME_STRING, "0"))
            hostnameIPAddressString = mySPR.getString(HOSTNAME_STRING, "0")!!
            binding.hostnameIpAddress.isEnabled = false
        }

        binding.hostnameIpAddressCheck.isChecked = mySPR.getBoolean(HOSTNAME_CHECK, true)

        if (mySPR.getInt(PORT_INT, 0) == 0) {
            binding.portText.setText("")
            binding.hostnameIpAddress.isEnabled = true
        } else {
            binding.portText.setText(mySPR.getInt(PORT_INT, 0).toString())
            binding.port.isEnabled = false
            portInt = mySPR.getInt(PORT_INT, 9001)
        }

        binding.portCheck.isChecked = mySPR.getBoolean(PORT_CHECK, true)

        val hostname = binding.hostnameIpAddressText.text.toString()
        val port = binding.portText.text.toString()
        if (hostname.isNotEmpty() && port.isNotEmpty()) {
            linkGeneration(hostname, port)
            binding.httpToggles[0].isEnabled = false
            binding.httpToggles[1].isEnabled = false
            binding.hostnameIpAddressCheck.isEnabled = false
            binding.portCheck.isEnabled = false
            binding.connectCheck.isEnabled = true
            binding.connectButton.text = getString(R.string.main_connect_button_2)
            binding.editButton.isEnabled = true
        }
    }

    private fun linkGeneration(hostname: String, port: String) {
        hostnameIPAddressString = hostname
        portInt = if (port.isNotEmpty()) {
            port.toInt()
        } else {
            9001
        }
        val prefix = if (requireView().findViewById<MaterialButton>(R.id.http_button).isChecked) {
            "http"
        } else {
            "https"
        }

        val portString = if (portInt != -1) ":$portInt" else ""
        link = "$prefix://$hostnameIPAddressString$portString"
        binding.linkView.text = link
    }

    private fun setConnectButtonClickListener(v: View) {
        binding.connectButton.setOnClickListener { view ->
            val tempHostname = binding.hostnameIpAddressText.text.toString().trim()
            val tempPort = binding.portText.text.toString().trim()
            if (tempHostname.isNotEmpty()) {
                binding.hostnameIpAddress.error = ""
                if (binding.linkView.text.toString().isEmpty()) {
                    linkGeneration(tempHostname, tempPort)

                    binding.httpToggles[0].isEnabled = false
                    binding.httpToggles[1].isEnabled = false
                    binding.hostnameIpAddress.isEnabled = false
                    binding.port.isEnabled = false
                    binding.hostnameIpAddressCheck.isEnabled = false
                    binding.portCheck.isEnabled = false
                    binding.connectCheck.isEnabled = true
                    binding.editButton.isEnabled = true

                    val httpMode = (binding.httpToggles[1] as MaterialButton).isChecked
                    editor.putBoolean(HTTPS_ACTIVATED, httpMode)

                    if (binding.hostnameIpAddressCheck.isChecked) {
                        editor.putString(HOSTNAME_STRING, hostnameIPAddressString)
                        editor.putBoolean(HOSTNAME_CHECK, true)
                    } else {
                        editor.putString(HOSTNAME_STRING, "")
                        editor.putBoolean(HOSTNAME_CHECK, false)
                    }

                    if (binding.portCheck.isChecked) {
                        editor.putBoolean(PORT_CHECK, true)

                        if (binding.portText.text.toString().trim().isNotEmpty()) {
                            editor.putInt(PORT_INT, portInt)
                        } else {
                            binding.portText.setText(9001.toString())
                            editor.putInt(PORT_INT, 9001)
                        }
                    } else {
                        binding.portText.setText(9001.toString())
                        editor.putInt(PORT_INT, 0)
                        editor.putBoolean(PORT_CHECK, false)
                    }

                    editor.putString(LINK, link).apply()

                    binding.connectButton.text = getString(R.string.main_connect_button_2)
                } else {
                    if (view != null) {
                        val imm = view.context.applicationContext.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                        imm.hideSoftInputFromWindow(view.windowToken, 0)
                    }
                    findNavController().navigate(R.id.nav_webview)

                    if (mySPR.getBoolean(CONNECTION_OVERVIEW_ENABLED, true)) {
                        storeAndShowConnection((binding.httpToggles[1] as MaterialButton).isChecked, tempHostname, tempPort.toInt())
                    }
                }
            } else {
                binding.hostnameIpAddress.error = getString(R.string.error_fill_out)
            }
        }
    }

    private fun storeAndShowConnection(httpsActivated: Boolean, hostname: String, port: Int) {
        val newConnection = Connection(httpsActivated, hostname, port)
        val connections = getListOfConnections(mySPR)
        if (!connections.contains(newConnection)) {
            if (connections.size >= 3) {
                connections.removeAt(0)
            }
            connections.add(newConnection)
            val newConnections = connections.joinToString(";")
            editor.putString(CONNECTIONS, newConnections).commit()
            showConnections(mySPR, connections, activity)
        }
    }

    private fun setEditButtonClickListener() {
        binding.editButton.setOnClickListener {
            binding.httpToggles[0].isEnabled = true
            binding.httpToggles[1].isEnabled = true
            binding.hostnameIpAddress.isEnabled = true
            binding.hostnameIpAddressCheck.isEnabled = true
            binding.port.isEnabled = true
            binding.portCheck.isEnabled = true
            binding.connectCheck.isEnabled = false
            binding.connectButton.text = getString(R.string.main_connect_button_1)
            binding.editButton.isEnabled = false
            binding.linkView.text = ""
        }
    }

    private fun setConnectCheckClickListener() {
        binding.connectCheck.setOnCheckedChangeListener { _, b ->
            editor.putBoolean(CONNECT_CHECK, b)
            editor.putBoolean(AUTO_START, b).apply()
        }
    }
}